### DISCUSSION TITLE
Color Module

### DISCUSSION THREAD

**Discussion by bonjorno7** [godotengine/godot-proposals#6892] [Engine Core]
https://github.com/godotengine/godot/pull/76976

I'm making a module for color space conversion and other such utilities, but don't know exactly what it needs.
Hopefully this discussion will help with that.

Currently it supports converting both ways between RGB and HSV / HSL / OK HSV / OK HSL.

**Comment by aaronfranke** [comment]
Some other things we could put in the module (and deprecate the corresponding core versions so we can delete later):

* Converting to/from integers (to/from rgba/argb/abgr 32/64, and also, we should replace the "hex" methods with from_ ones)
* Named constants (does the core engine really need to know the value of "lavender_blush" or "medium_aquamarine"?)
    * The named constants could go in a class called `Colors` instead of the rest which would be on `ColorUtil` or similar (both still in the module). This would match [the `Colors` static class in C#](https://github.com/godotengine/godot/blob/master/modules/mono/glue/GodotSharp/GodotSharp/Core/Colors.cs).

**Reply by bonjorno7** [reply]
Should this Colors class be part of the color_utils module?

**Reply by bonjorno7** [reply]
I've added rgba32/64 functions and a Colors class.
Is there a reason to use a dictionary instead of just a bunch of consts?

**Reply by aaronfranke** [reply]
It's so that you can do `Color.from_string("red")` in addition to `Color.RED`. So with this refactoring I'd imagine we could do `ColorUtil.from_string("red")` and `Colors.RED`. Note that `from_string` also allows creating from a hex string.

**Reply by bonjorno7** [reply]
Should this function be in Colors or ColorUtils?
Personally I'd put it in Colors and call it get, because it's short and simple.
If it's in ColorUtils I'd call it from_name or something, because it's not just any string, but a name from a specific set.

**Reply by aaronfranke** [reply]
Well the existing function works for both `"red"` and `"#a1b2c3"`, which I think is worth keeping. But yeah we could put just the from_name part in `Colors` (existing `Color::named`) and then `ColorUtil(s)::from_string` just calls this:

```cpp
Color ColorUtil::from_string(const String &p_string, const Color &p_default) {
	if (html_is_valid(p_string)) {
		return from_html(p_string); // ColorUtil::from_html
	} else {
		return Colors::from_name(p_string, p_default);
	}
}
```

**Reply by bonjorno7** [reply]
Oh I didn't realize the existing from_string supported hex too.
In that case I agree it makes sense to put it in ColorUtils as from_string and replicate the existing behavior.
Preferably with named in Colors to avoid clutter in ColorUtils yeah.

**Reply by bonjorno7** [reply]
Currently trying to figure out how to add constants to a GD class.
I'll go to bed now and continue in the morning.

**Reply by bonjorno7** [reply]
It may not be possible to have non-integer constants in a class inside a module, because all such constants are registered in _VariantCall, in the engine core.

**Reply by bonjorno7** [reply]
Exposing add_variant_constant doesn't help, because you can only add constants to supported variant types, not just any class.

**Reply by aaronfranke** [reply]
That's a good point, we may need to add this feature to the engine before we can move the constants.

**Reply by bonjorno7** [reply]
Is that something quick I should go do (in a separate PR)?
Or would this be difficult, perhaps better left for the experts?

**Reply by bonjorno7** [reply]
So far I've edited 14 files and I don't think I'm making much progress towards actually making it work.
Here's my attempt: https://github.com/bonjorno7/godot/tree/variant-constant

**Reply by bonjorno7** [reply]
Alright it seems I've got it working: https://github.com/godotengine/godot/pull/77293

**Reply by bonjorno7** [reply]
That's a huge feature though, and I think we should be able to add the color_utils module without it for now.
